File: basic_dsa_three_hundred_eighteen.java
/**
 * Basic DSA Problems - Middle of the Linked List
 * Find the middle node of a linked list
 */
public class basic_dsa_three_hundred_eighteen {

    // Problem: Head of linked list. Return middle node. If two middles, return second.

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    /**
     * Fast and slow pointers: slow moves 1, fast moves 2. When fast reaches end, slow is middle.
     * Time O(n), Space O(1)
     */
    public static ListNode middleNode(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Middle of the Linked List");
        System.out.println("=".repeat(60));

        ListNode a = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, null)))));
        System.out.println("[1,2,3,4,5] -> " + middleNode(a).val);

        ListNode b = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, new ListNode(6, null))))));
        System.out.println("[1,2,3,4,5,6] -> " + middleNode(b).val);
    }
}

File: basic_dsa_three_hundred_eighteen.cpp
/**
 * Basic DSA - Odd Even Linked List (LeetCode 328)
 * Group odd-indexed nodes then even-indexed nodes (1-based).
 */
#include <iostream>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* oddEvenList(ListNode* head) {
    if (!head || !head->next) return head;
    ListNode *odd = head, *even = head->next, *evenHead = even;
    while (even && even->next) {
        odd->next = even->next;
        odd = odd->next;
        even->next = odd->next;
        even = even->next;
    }
    odd->next = evenHead;
    return head;
}

void printList(ListNode* h) {
    while (h) {
        std::cout << h->val << (h->next ? "->" : "");
        h = h->next;
    }
    std::cout << "\n";
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Odd Even Linked List\n";
    std::cout << std::string(60, '=') << "\n";

    ListNode* a = new ListNode(1);
    a->next = new ListNode(2);
    a->next->next = new ListNode(3);
    a->next->next->next = new ListNode(4);
    a->next->next->next->next = new ListNode(5);
    printList(oddEvenList(a));

    ListNode* b = new ListNode(2);
    b->next = new ListNode(1);
    b->next->next = new ListNode(3);
    b->next->next->next = new ListNode(5);
    b->next->next->next->next = new ListNode(6);
    b->next->next->next->next->next = new ListNode(4);
    b->next->next->next->next->next->next = new ListNode(7);
    printList(oddEvenList(b));
}
